java-squid/src/main/java/org/sonar/java/se/SymbolicValueRelation.java

```diff
@@ -34,73 +38,154 @@ protected SymbolicValueRelation(SymbolicValue v1, SymbolicValue v2) {
+    List<SymbolicValueRelation> result = new ArrayList<>();
+    for (SymbolicValueRelation relation : relations) {
+      SymbolicValueRelation combined = null;
+      if (v2.equals(relation.v1) && !v1.equals(relation.v2)) {
+        combined = relation.combinedAfter(this);
+      } else if (v2.equals(relation.v2) && !v1.equals(relation.v2)) {
+        combined = relation.symmetric().combinedAfter(this);
+      }
+      if (combined != null && !usedRelations.contains(combined)) {
+        result.add(combined);
+      }
+    }
+    return result;
   }
 
   @Override
   public String toString() {
     StringBuilder buffer = new StringBuilder();
-    buffer.append('(');
     buffer.append(v1);
     buffer.append(' ');
     buffer.append(getOperand());
     buffer.append(' ');
     buffer.append(v2);
-    buffer.append(')');
     return buffer.toString();
   }
 
   protected abstract String getOperand();
 
+  /**
+   * @return a new relation, the negation of the receiver
+   */
   protected abstract SymbolicValueRelation inverse();
 
   /**
+   * @return a new relation, equivalent to the receiver with inverted parameters
+   */
+  protected abstract SymbolicValueRelation symmetric();
+
+  /**
    * @param a relation between symbolic values
    * @return <ul>
-   * <li>TRUE  if the supplied relation implies that the receiver is true</li>
-   * <li>FALSE if the supplied relation implies that the receiver is false</li>
+   * <li>TRUE  if the receiver implies that the supplied relation is true</li>
+   * <li>FALSE if the receiver implies that the supplied relation is false</li>
    * <li>null  otherwise</li>
    * </ul>
    */
-  protected abstract Boolean implies(SymbolicValueRelation relation);
+  @CheckForNull
+  public Boolean implies(SymbolicValueRelation relation) {
+    if (v1.equals(relation.v1) && v2.equals(relation.v2)) {
+      return relation.isImpliedBy(this);
+    } else if (v1.equals(relation.v2) && v2.equals(relation.v1)) {
+      return relation.isImpliedBy(symmetric());
+    }
+    return null;
+  }
 
-  /**
-   * Returns a new relation to be checked against a new set of hypotheses without the supplied relation.
-   * The precondition is that none of the known hypotheses yields direct information on the receiver (the checked relation).
-   * This result is a combination of the receiver (currently checked relation against the set of hypotheses) and the supplied relation.
-   * If the receiver combined with the supplied relation does not yield new information, null is returned.
-   *
-   * @param relation, belonging to the set of known hypotheses
-   * @return a SymbolicValueRelation or null
-   */
-  @Nullable
-  protected abstract SymbolicValueRelation underHypothesis(SymbolicValueRelation relation);
+  @CheckForNull
+  protected abstract Boolean isImpliedBy(SymbolicValueRelation relation);
+
+  @CheckForNull
+  protected abstract Boolean impliesEqual();
```

Comment:
I'm really not sure that having a method `implies` for each other `symbolicValueRelation` is a good approach... When/if we want to add a relation that as nothing to do with the existing, we will have to implements all these methods for nothing.